programs/perpetuals/src/instructions/set_custom_oracle_price_permissionless.rs

```diff
@@ -0,0 +1,128 @@
+//! SetCustomOraclePricePermissionless instruction handler
+
+use {
+    crate::{
+        error::PerpetualsError,
+        state::{custody::Custody, oracle::CustomOracle, perpetuals::Perpetuals, pool::Pool},
+    },
+    anchor_lang::prelude::*,
+    solana_program::{
+        ed25519_program::ID as ED25519_ID,
+        instruction::Instruction,
+        sysvar::instructions::{load_instruction_at_checked, ID as IX_ID},
+    },
+};
+
+#[derive(Accounts)]
+pub struct SetCustomOraclePricePermissionless<'info> {
+    #[account(mut)]
+    pub user: Signer<'info>,
+
+    #[account(
+        seeds = [b"perpetuals"],
+        bump = perpetuals.perpetuals_bump
+    )]
+    pub perpetuals: Box<Account<'info, Perpetuals>>,
+
+    #[account(
+        seeds = [b"pool",
+                 pool.name.as_bytes()],
+        bump = pool.bump
+    )]
+    pub pool: Box<Account<'info, Pool>>,
+
+    #[account(
+        seeds = [b"custody",
+                 pool.key().as_ref(),
+                 custody.mint.as_ref()],
+        bump = custody.bump
+    )]
+    pub custody: Box<Account<'info, Custody>>,
+
+    #[account(
+        // Custom oracle must first be initialized by authority before permissionless updates.
+        mut,
+        seeds = [b"oracle_account",
+                 pool.key().as_ref(),
+                 custody.mint.as_ref()],
+        bump
+    )]
+    pub oracle_account: Box<Account<'info, CustomOracle>>,
+
+    /// CHECK: Needed for ed25519 signature verification, to inspect all instructions in this transaction.
+    #[account(address = IX_ID)]
+    pub ix_sysvar: AccountInfo<'info>,
+}
+
+#[derive(AnchorSerialize, AnchorDeserialize, Copy, Clone)]
+pub struct SetCustomOraclePricePermissionlessParams {
+    pub price: u64,
+    pub expo: i32,
+    pub conf: u64,
+    pub ema: u64,
+    pub publish_time: i64,
+}
+
+pub fn set_custom_oracle_price_permissionless<'info>(
+    ctx: Context<'_, '_, '_, 'info, SetCustomOraclePricePermissionless<'info>>,
+    params: &SetCustomOraclePricePermissionlessParams,
+) -> Result<()> {
+    // If the incoming publish time is behind or equal to the current, then no-op.
+    if params.publish_time <= ctx.accounts.oracle_account.publish_time {
+        msg!("Publish time is behind the current, so CustomOracle price did not update.");
```

Comment:
This log message doesn't sound well. Lets reword it or remove entirely

Reply:
ahh yeah sorry about that!  how is the updated message?